Add initial entitlement policy parsing

jdconrad opened the pull request:
This change adds entitlement policy parsing with the following design:
1. YAML file for readability and re-use of our x-content parsers
2. hierarchical structure to group entitlements under a single scope
3. no general entitlements without a scope or for the entire project

This is just the parsing logic. I would like early feedback before moving onto the immutable data structures we store the entitlements in for comparison.

Design decisions:
1. The parsing logic is contained inside a set of builders. I anticipate the builders will have another a method added to "build" and validate the parsed values into immutable objects after the parsing is completed. This will allow us to also convert module names into module objects once they are available.
2. I broke the builders into different policy, scope, and entitlement. This makes it easier to add scopes and entitlements rather than have a single giant builder.
3. I opted to do the x-content parsing by hand rather than use `ConstructingObjectParser` as this allows for better error messages and is often more intuitive in debugging with breakpoints if necessary.
4. I added a `PolicyParserException` to give consistent, detailed error messages throughout.

jdconrad committed:
start of policy parser

jdconrad committed:
initial parser for module scope

jdconrad committed:
split builders into pieces and clean up error messaging

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

mark-vieira commented on libs/entitlement-runtime/build.gradle:
```diff
@@ -11,12 +11,12 @@ apply plugin: 'elasticsearch.publish'
-tasks.named('forbiddenApisMain').configure {
-  replaceSignatureFiles 'jdk-signatures'
+  testImplementation(project(":test:framework")) {
+    exclude group: 'org.elasticsearch', module: 'elasticsearch-entitlement-runtime'
```
Why would `test-framework` bring in the entitlement-runtime project?

jdconrad replied:
I didn't realize this project had been moved from libs to distribution. I'm working on fixing everything now when merging with main.

mark-vieira replied:
Yeah, I think you can just remove this exclude as it's unnecessary.

jdconrad replied:
I think this is fixed with the main merge conflicts resolved.

prdoyle commented on libs/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/policy/EntitlementBuilder.java:
```diff
@@ -0,0 +1,29 @@
+        this.policyParser = policyParser;
+    }
+
+    protected abstract void parseEntitlement() throws IOException;
```
What is the contract of this method? From the implementations, it seems to use the parser to validate the structure, but doesn't actually have any other side effects or return anything?

Is the idea that this will ultimately return whatever data structure we end up with?

jdconrad replied:
I was thinking there would be two passes. One for parsing and one for building. The parsing one would validate things like is the structure correct and are these values strings, etc. The building would validate that the scopes exist and permissions have valid paths, etc then turn them into immutable structures that will be used for comparison. Open here to a single pass if we think that's better. Either way we will need to do a string lookup at some point to get the module object for comparison.

ldematte replied:
I see. It makes sense, if you think there will be a **public** build method.
Maybe to make this clear these classes could be internal (package private?)

prdoyle replied:
That sounds like an interesting separation, between syntax and semantics. Phase 1: make sure you can tell what they're asking for; phase 2: make sure they're asking for something reasonable.

jdconrad replied:
I'm not 100% sure this is the right approach, yet, but have to start somewhere. Either way parsing code will be useful.

prdoyle commented on libs/entitlement-runtime/src/test/java/org/elasticsearch/entitlement/policy/PolicyBuilderTests.java:
```diff
@@ -0,0 +1,19 @@
+public class PolicyBuilderTests extends ESTestCase {
+
+    public void testPolicyBuilder() {
+        new PolicyBuilder("test-policy.yaml", PolicyBuilderTests.class.getResourceAsStream("test-policy.yaml")).buildPolicy();
```
This needs some sad-path tests too, with various kinds of invalid YAML structures.

jdconrad replied:
Definitely!

prdoyle replied:
In the past I have had some success with a tool called pitest to inspire unit test cases based on the sorts of bugs they would detect. It can be tricky to set up though.

prdoyle commented on libs/entitlement-runtime/src/test/resources/org/elasticsearch/entitlement/policy/test-policy.yaml:
```diff
@@ -0,0 +1,9 @@
+policy:
+  - module:
+      name: entitlement-test
+      entitlements:
```
How do you envision this structure extending to support class-level grants?

jdconrad replied:
I envisioned having sub objects of module for package and then for class with their own set of entitlements for those specific objects. @rjernst and I had chatted about having all scoping be at the top-level where scope was separate by a colon, but I thought the trade off of having them be sub objects made it easier for users to see what permissions belong where. I'm really open to changing this either way.

prdoyle replied:
Oh I think I get it. This is the hierarchical structure you mentioned. I like it!

jdconrad replied:
I think it makes sense. We can certainly make changes to simplify or move structure around if we want to.

jdconrad committed:
Merge branch 'main' into policy

ldematte commented on distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/policy/PolicyBuilder.java:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+    public void buildPolicy() {
```
IMO this should be named consistently with the others (parsePolicy?) and be package private. So it's still accessible by tests, but it's clear that we will use `build()` instead (to be added).

ldematte reviewed (commented):
Looks like a very good start, thanks!
Looking forward to see the final shape, with the Builder -> immutable objects.
I have a slight concern with naming: they are not really what you would expect from a builder, are they? The only one vaguely resembling a "traditional" java builder is ModuleScopeBuilder.
Why don't we call them Parser or Reader?

prdoyle commented on distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/policy/FileEntitlementBuilder.java:
```diff
@@ -0,0 +1,120 @@
+        super(policyName, scopeName, policyParser);
+    }
+
+    protected void parseEntitlement() throws IOException {
```
NGL, I didn't read the implementation here in detail. This seems like a fairly low-stakes situation; I assume you did smart things in here, and if we find out something isn't right, we can just fix it. An extensive test suite is probably better for demonstrating that this sort of method is right, as opposed to careful review.

jdconrad replied:
I understand :) . Tests should cover this kind of thing typically.

prdoyle commented on distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/policy/PolicyBuilder.java:
```diff
@@ -0,0 +1,110 @@
+                throw newPolicyParserException(policyParser.getTokenLocation(), policyName, "expected closing object");
+            }
+        } catch (IOException ioe) {
+            throw new UncheckedIOException(ioe);
```
Given that this is using an `InputStream`, should the method be declared `throws IOException`?

jdconrad replied:
Changed.

prdoyle commented on distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/policy/PolicyParserException.java:
```diff
@@ -0,0 +1,89 @@
+
+import org.elasticsearch.xcontent.XContentLocation;
+
+public class PolicyParserException extends RuntimeException {
```
I'm not sure what Elastic's philosophy is on runtime exceptions. This strikes me as a situation where this feature (the policy parser) will be used in very few places--probably just one--and in that place, the caller is going to need to catch this exception and do something about it anyway. That means the drawbacks of checked exceptions are insignificant, and making it a checked exception might help remind future devs of the sharp edges here.

jdconrad replied:
Most of our other parse exceptions are also runtime exceptions. I think I'd prefer to go with that precedent for now. If a policy fails, ES will not start anyway, so is should be pretty obvious to anyone adding code around this what happened.

prdoyle commented on distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/policy/EntitlementBuilder.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.io.IOException;
+
+public abstract class EntitlementBuilder {
```
I wonder, given the multi-phase parse-validate-build approach you mentioned in the comments... should this be called `EntitlementParser`?

ldematte replied:
++ (I suggested the same, Parser or Reader)

jdconrad replied:
I renamed everything to parser. I had considered these builders for the validation/build step, but maybe what makes the most sense is to separate out the builders completely from the parsers. So it would end up being parser -> builder -> immutable policy object(s). If that's the case I'll have the parsers return builder objects. WDYT?

prdoyle replied:
FWIW my hunch is that the extra step will pay off. There's a tendency to make parsers do a lot, but unless they're very simple, parsers are best left just determining what the text _says_ rather than trying to validate that it makes sense, or trying to "execute" it by building some sort of efficient data structures.

It makes sense to parse into some convenient form, validate that, and then once we have all the validated grants, run a build step that generates a highly efficient data structure for entitlement checks.

It's not super clear though, because it's possible our situation will be simple enough that this distinction isn't pulling its weight, and the parser could just generate efficient data structures directly. 🤷

jdconrad committed:
Merge branch 'main' into policy

jdconrad commented on distribution/tools/entitlement-runtime/build.gradle:
```diff
@@ -11,16 +11,12 @@ apply plugin: 'elasticsearch.publish'
 
 dependencies {
   compileOnly project(':libs:elasticsearch-core') // For @SuppressForbidden
+  compileOnly project(":libs:elasticsearch-x-content") // for parsing policy files
```
I'm open to moving it for sure, but for now this still seems like the best starting point to me.

jdconrad committed:
update based on pr comments

jdconrad commented:
@ldematte @prdoyle Thanks for the reviews! I renamed everything to parser. I'm working on adding failure tests. Please let me know if there's anything else you would expect from this PR as I think this is a good start pointing to iterate from.

jdconrad committed:
spotless

ldematte commented:
@jdconrad I think this is a good start pointing 
At this stage, I'm not sure if we need a separate set of builders too; I think you can start maybe with defining the immutable objects, and then see if you need extra steps to build them from the parsers. That will tell you if a set of separate builders is needed of if it is a non necessary complication.

jdconrad committed:
add failure tests class

jdconrad committed:
Merge branch 'main' into policy

jdconrad committed:
move packages

jdconrad committed:
Merge branch 'main' into policy

jdconrad committed:
update parser to remove boilerplate

jdconrad commented:
@ldematte @prdoyle I have updated this PR based on some feedback about removing future boilerplate. I have modified the design based on our discussions with the following significant changes:
1. The `PolicyParser` is now a single class that is much simpler.
2. All `Entitlement` classes are now expected to have a single constructor, and if available to plugins (or other external policy files) expected to annotate that constructor with `ExternalEntitlement`. This allows the parser to reflectively find available entitlements meaning we do not need any additional parsing as entitlements are potentially added. (Thus, way less boilerplate.)
3. Each entitlement is expected to do its own validation as part of its constructor. This allows the parser to stay simple, and validation to occur for internally built entitlements such as we expect to do with system exit.
4. This PR contains the very basic immutable classes `Policy`, `Scope`, and `Entitlement`. This uses `FileEntitlement` as the initial example with several tests build around it.
5. My expectation is that the immutable classes will be updated either lazily or after plugin classes are available in optimized versions, but that is certainly for a follow up.
6. `Scope` can be extended to have a child scope if we need it, but for now the assumption is that the only supported scope is modules.

rjernst commented on distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/policy/FileEntitlement.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.List;
+import java.util.Objects;
+
+public class FileEntitlement implements Entitlement {
```
Perhaps this could be a record? That would remove more boilerplate.  But that could be in a followup.

jdconrad replied:
I will leave this as a follow up because I'm not sure we don't want to add some kind of entitlement comparison in each one where this one would take in a path and an action to see if it's permitted. (Not a new object.)

rjernst commented on distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/policy/Entitlement.java:
```diff
@@ -0,0 +1,14 @@
+
+package org.elasticsearch.entitlement.runtime.policy;
+
+public interface Entitlement {
```
Some javadocs would be nice for other devs to know this is a marker interface

jdconrad replied:
Added.

rjernst commented on distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/policy/ExternalEntitlement.java:
```diff
@@ -0,0 +1,19 @@
+@Retention(RetentionPolicy.RUNTIME)
+public @interface ExternalEntitlement {
+
+    String[] parameterNames() default {};
```
Similar here, javadocs to explain why this parameterNames exists at all

jdconrad replied:
Added.

rjernst commented on distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/policy/ExternalEntitlement.java:
```diff
@@ -0,0 +1,19 @@
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+public @interface ExternalEntitlement {
```
javadoc?

prdoyle replied:
Also `@Target`.

jdconrad replied:
Added both documentation and an @Target for constructor.

rjernst commented on distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/policy/FileEntitlement.java:
```diff
@@ -0,0 +1,64 @@
+    public static final int READ_ACTION = 0x1;
+    public static final int WRITE_ACTION = 0x2;
+
+    private final String path;
```
We'll need to access these, so they should be public or have accessors?

jdconrad replied:
This depends on the future design if we end up doing entitlement checks as part of each entitlement class or not. I would prefer to change this in a follow up to match the design we decide on moving forward.

rjernst commented on distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/policy/PolicyParser.java:
```diff
@@ -0,0 +1,215 @@
+            );
+        } catch (ClassNotFoundException cnfe) {
+            throw newPolicyParserException(
+                policyParser.getTokenLocation(),
```
nit: The token location and policy name seem to be the same in all calls to these helper exception methods, could those be removed as args?

jdconrad replied:
I cleaned this up with some additional helper methods.

rjernst commented on distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/policy/Scope.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.List;
+import java.util.Objects;
+
+public class Scope {
```
Could this be a record?

jdconrad replied:
It depends on how we check entitlements moving forward. If they all become part of the top-level `Policy` class then I'm happy to change it then.

rjernst commented on distribution/tools/entitlement-runtime/src/test/java/org/elasticsearch/entitlement/runtime/policy/PolicyParserFailureTests.java:
```diff
@@ -0,0 +1,149 @@
+
+    public void testEntitlementExtraneousParameter() throws IOException {
+        try (XContentBuilder builder = YamlXContent.contentBuilder()) {
+            builder.startObject();
```
It would be much easier to read these tests if they used multi line strings, so the file content is more understandable. I know we have used builders in tests in the past for much of xcontent, indexing, etc, but in those cases I think we care about varying the underlying xcontent type. Here we only care about yaml.

jdconrad replied:
Changed.

rjernst commented on distribution/tools/entitlement-runtime/src/test/resources/org/elasticsearch/entitlement/runtime/policy/test-policy.yaml:
```diff
@@ -0,0 +1,7 @@
+entitlement-module-1:
```
nit: maybe make this "example-module-name", with the `1` it looks like this is an arbitrary id

jdconrad replied:
Changed.

rjernst reviewed (commented):
I have a few more suggestions, but this looks pretty good to me!

ldematte reviewed (approved):
Looks really good to me!
I agree with Ryan that more javadoc would be very nice (possibly copy/pasting some of the explanations from comments on this PR too). 
Other than that, it looks good to me; I agree that more things (adding other entitlements, the "final" immutable classes, child scopes) can be added later, when needed.

jdconrad committed:
Merge branch 'main' into policy

jdconrad committed:
changes based on pr feedback

jdconrad commented:
@ldematte @prdoyle @rjernst Thank you all for the reviews! I've updated the PR based on your latest feedback, and will merge as soon as CI passes. If there's anything I missed, I'm happy to change in a follow up.
